OpenKeychain/src/main/java/org/sufficientlysecure/keychain/pgp/UncachedKeyRing.java

```diff
@@ -157,6 +159,50 @@ public static UncachedKeyRing decodeFromData(byte[] data)
 
     }
 
+    public static UncachedKeyRing decodeFromData(
+            byte[] data, String expectedFingerprint, String expectedKeyId)
+            throws PgpGeneralException, IOException {
+
+        IteratorWithIOThrow<UncachedKeyRing> parsed = fromStream(new ByteArrayInputStream(data));
+
+        if ( ! parsed.hasNext()) {
+            throw new PgpGeneralException("Object not recognized as PGPKeyRing!");
+        }
+
+        UncachedKeyRing ringToAccept = null;
+
+        while (parsed.hasNext()) {
+            UncachedKeyRing ring = parsed.next();
+            Iterator<UncachedPublicKey> publicKeys = ring.getPublicKeys();
+
+            while (publicKeys.hasNext()) {
+                UncachedPublicKey publicKey = publicKeys.next();
+
+                boolean publicKeyMatches = (
+                        KeyFormattingUtils.convertFingerprintToHex(publicKey.getFingerprint())
+                                .equalsIgnoreCase(expectedFingerprint)
+                        || TextUtils.equals(Long.toString(publicKey.getKeyId()), expectedKeyId));
```

Comment:
seems better to turn expectedKeyId into a Long earlier, then compare if non-null